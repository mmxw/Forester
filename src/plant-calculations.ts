import {
  Plant,
  PlantKind,
  PlantKindId,
  PlantState,
  UIPlant,
  WaterFrequency,
} from './types';

// TODO: better date math and display https://github.com/mmxw/Forester/issues/12
const ONE_DAY = 1000 * 60 * 60 * 24;
const ONE_MONTH = ONE_DAY * 30;
const ONE_WEEK = ONE_DAY * 7;

export function plantToUIPlant(
  {contact, lastWatered, waterFrequency, plantKindId}: Plant,
  plantKinds: PlantKind[],
  now: Date,
): UIPlant {
  const plantKind = findPlantKind(plantKinds, plantKindId);
  const [state, nextWatering] = calcPlantStateAndNextWatering(
    now,
    lastWatered,
    waterFrequency,
  );
  const appearance = plantKind.appearances[state];

  return {
    name: contact.name,
    appearance,
    lastWatered,
    nextWatering,
    waterFrequency,
    state,
  };
}

function calcPlantStateAndNextWatering(
  now: Date,
  lastWatered: Date,
  waterFrequency: WaterFrequency,
): [PlantState, Date] {
  const frequencyMillis = waterFrequencyToMilliseconds(waterFrequency);
  const diff = now.valueOf() - lastWatered.valueOf();
  if (diff > frequencyMillis * 2) {
    return ['departed', now];
  }
  if (diff > frequencyMillis) {
    return ['droopy', now];
  }
  return ['happy', new Date(now.valueOf() + diff)];
}

function waterFrequencyToMilliseconds({number, unit}: WaterFrequency): number {
  switch (unit) {
    case 'days':
      return number * ONE_DAY;
    case 'months':
      return number * ONE_MONTH;
    case 'weeks':
      return number * ONE_WEEK;
  }
  // eslint-disable-next-line
  const _exhaustivenessCheck: never = unit;
}

function findPlantKind(
  plantKinds: PlantKind[],
  plantKindId: PlantKindId,
): PlantKind {
  const kind = plantKinds.find((k) => k.id === plantKindId);
  if (!kind) {
    // todo: better error handling for assertions https://github.com/mmxw/Forester/issues/10
    throw Error(`invalid data! could not find plant kind id ${plantKindId}`);
  }
  return kind;
}
