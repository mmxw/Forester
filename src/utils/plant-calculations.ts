/**
 * This file has utilities for converting from our internal data format
 * to a format that is closer to what the UI needs.
 *
 * For example, the code in this file calcuates what plant icon to show
 * based on watering frequency, species, last watering, etc.
 */
import {ONE_DAY, ONE_MONTH, ONE_WEEK} from './date';
import {
  Plant,
  Species,
  SpeciesId,
  PlantState,
  UIPlant,
  WaterFrequency,
} from './types';

export function plantToUIPlant(
  {plantId, contact, lastWatered, waterFrequency, speciesId}: Plant,
  speciesArr: Species[],
  now: Date,
): UIPlant {
  const species = findSpecies(speciesArr, speciesId);
  const [state, nextWatering] = calcPlantStateAndNextWatering(
    now,
    lastWatered,
    waterFrequency,
  );
  const appearance = species.appearances[state];

  return {
    plantId,
    name: contact.name,
    appearance,
    lastWatered,
    nextWatering,
    waterFrequency,
    state,
  };
}

function calcPlantStateAndNextWatering(
  now: Date,
  lastWatered: Date,
  waterFrequency: WaterFrequency,
): [PlantState, Date] {
  const frequencyMillis = waterFrequencyToMilliseconds(waterFrequency);
  const diff = now.valueOf() - lastWatered.valueOf();
  if (diff > frequencyMillis * 2) {
    return ['departed', now];
  }
  if (diff > frequencyMillis) {
    return ['droopy', now];
  }
  return ['happy', new Date(lastWatered.valueOf() + frequencyMillis)];
}

function waterFrequencyToMilliseconds({number, unit}: WaterFrequency): number {
  switch (unit) {
    case 'days':
      return number * ONE_DAY;
    case 'months':
      return number * ONE_MONTH;
    case 'weeks':
      return number * ONE_WEEK;
  }
}

function findSpecies(speciesArr: Species[], speciesId: SpeciesId): Species {
  const species = speciesArr.find((k) => k.id === speciesId);
  /* istanbul ignore next */
  if (!species) {
    // todo: better error handling for assertions https://github.com/mmxw/Forester/issues/10
    throw Error(`invalid data! could not find plant species id ${speciesId}`);
  }
  return species;
}
