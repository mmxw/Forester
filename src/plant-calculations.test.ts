import {plantToUIPlant} from './plant-calculations';
import type {Plant, PlantId, PlantKind, PlantKindId} from './types';

test('plantToUIPlant', () => {
  const now = new Date('January 15, 2021');
  const plants: Plant[] = [
    {
      lastWatered: new Date('January 10, 2021'),
      plantKindId: 'plantKind1' as PlantKindId,
      waterFrequency: {
        number: 1,
        unit: 'weeks',
      },
      plantId: 'p1' as PlantId,
      contact: {
        recordId: 'r1',
        emails: [],
        name: 'person 1',
        phones: [],
        postalAddresses: [],
      },
    },
    {
      lastWatered: new Date('January 13, 2021'),
      plantKindId: 'plantKind2' as PlantKindId,
      waterFrequency: {
        number: 1,
        unit: 'days',
      },
      plantId: 'p2' as PlantId,
      contact: {
        recordId: 'r2',
        emails: [],
        name: 'person 2',
        phones: [],
        postalAddresses: [],
      },
    },
    {
      lastWatered: new Date('January 12, 2021'),
      plantKindId: 'plantKind2' as PlantKindId,
      waterFrequency: {
        number: 1,
        unit: 'days',
      },
      plantId: 'p2' as PlantId,
      contact: {
        recordId: 'r2',
        emails: [],
        name: 'person 2',
        phones: [],
        postalAddresses: [],
      },
    },
  ];

  const plantKinds: PlantKind[] = [
    {
      id: 'plantKind1' as PlantKindId,
      name: 'plant kind 1',
      appearances: {
        happy: {
          emoji: '🤣',
          kind: 'emoji',
        },
        droopy: {
          emoji: '😢',
          kind: 'emoji',
        },
        departed: {
          emoji: '🤢',
          kind: 'emoji',
        },
      },
    },
    {
      id: 'plantKind2' as PlantKindId,
      name: 'face kind 2',
      appearances: {
        happy: {
          emoji: '👍',
          kind: 'emoji',
        },
        droopy: {
          emoji: '🤦‍♀️',
          kind: 'emoji',
        },
        departed: {
          emoji: '👎',
          kind: 'emoji',
        },
      },
    },
  ];

  expect(
    plants.map((plant) => plantToUIPlant(plant, plantKinds, now)),
  ).toMatchSnapshot();
});
