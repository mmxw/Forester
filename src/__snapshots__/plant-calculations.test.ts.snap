// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plantToUIPlant 1`] = `
Array [
  Object {
    "appearance": Object {
      "emoji": "ü§£",
      "kind": "emoji",
    },
    "lastWatered": 2021-01-10T00:00:00.000Z,
    "name": "person 1",
    "nextWatering": 2021-01-20T00:00:00.000Z,
    "state": "happy",
    "waterFrequency": Object {
      "number": 1,
      "unit": "weeks",
    },
  },
  Object {
    "appearance": Object {
      "emoji": "ü§¶‚Äç‚ôÄÔ∏è",
      "kind": "emoji",
    },
    "lastWatered": 2021-01-13T00:00:00.000Z,
    "name": "person 2",
    "nextWatering": 2021-01-15T00:00:00.000Z,
    "state": "droopy",
    "waterFrequency": Object {
      "number": 1,
      "unit": "days",
    },
  },
  Object {
    "appearance": Object {
      "emoji": "üëé",
      "kind": "emoji",
    },
    "lastWatered": 2021-01-12T00:00:00.000Z,
    "name": "person 2",
    "nextWatering": 2021-01-15T00:00:00.000Z,
    "state": "departed",
    "waterFrequency": Object {
      "number": 1,
      "unit": "days",
    },
  },
]
`;
